File: table-engine/src/structured-refs.ts
/**
 * Structured Reference parsing, resolution, adjustment, and formatting.
 *
 * Heavy computation delegates to Rust/WASM via compute-core.
 *
 * Supports Excel-style structured references:
 *   Table1[Column1]              - single column data
 *   Table1[@Column1]             - this row, single column
 *   Table1[#Headers]             - entire header row
 *   Table1[#Data]                - entire data area
 *   Table1[#Totals]              - entire totals row
 *   Table1[#All]                 - entire table
 *   Table1[#This Row]            - this row of data
 *   Table1[[#Headers],[Column1]] - header cell of Column1
 *   Table1[[#Totals],[Col1]:[Col3]] - totals row, columns 1-3
 *   Table1[[Col1]:[Col3]]        - data range across columns 1-3
 */

import type { CellRange, StructuredRef, Table, TableStructureChange } from './types';

import { getWasm } from './wasm-backend';

// =============================================================================
// Parsing (delegates to WASM)
// =============================================================================

/**
 * Parse a structured reference string into a StructuredRef.
 *
 * Returns null for invalid input (never throws).
 */
export function parseStructuredRef(text: string): StructuredRef | null {
  const result = getWasm().table_parse_structured_ref(text);
  return (result ?? null) as StructuredRef | null;
}

// =============================================================================
// Resolution (delegates to WASM)
// =============================================================================

/**
 * Resolve a structured reference to concrete grid ranges.
 *
 * May produce multiple ranges for union refs (e.g., headers + column).
 * Returns empty array if the reference cannot be resolved.
 *
 * @param ref - Parsed structured reference
 * @param table - Table definition
 * @param currentRow - Current formula row for #This Row resolution (optional)
 */
export function resolveStructuredRef(
  ref: StructuredRef,
  table: Table,
  currentRow?: number,
): readonly CellRange[] {
  const result = getWasm().table_resolve_structured_ref(ref, table, currentRow ?? null);
  return result as CellRange[];
}

// =============================================================================
// Adjustment (delegates to WASM)
// =============================================================================

/**
 * Adjust a structured reference in response to a table structure change.
 *
 * Returns a new StructuredRef (or the same one if no change needed).
 */
export function adjustStructuredRef(
  ref: StructuredRef,
  change: TableStructureChange,
): StructuredRef {
  return getWasm().table_adjust_structured_ref(ref, change) as StructuredRef;
}

// =============================================================================
// Formatting (delegates to WASM)
// =============================================================================

/**
 * Format a StructuredRef back to its string representation.
 *
 * Roundtrip: parseStructuredRef(formatStructuredRef(ref)) should produce
 * an equivalent StructuredRef.
 */
export function formatStructuredRef(ref: StructuredRef): string {
  return getWasm().table_format_structured_ref(ref) as string;
}

File: compute/core/crates/compute-table/src/structured_refs.rs
//! Structured Reference parsing, resolution, adjustment, and formatting.
//!
//! Pure, stateless functions. No DOM, no Yjs, no React, no XState.
//! Ported from `table-engine/src/structured-refs.ts`.
//!
//! Supports Excel-style structured references:
//!   Table1[Column1]              — single column data
//!   Table1[@Column1]             — this row, single column
//!   Table1[#Headers]             — entire header row
//!   Table1[#Data]                — entire data area
//!   Table1[#Totals]              — entire totals row
//!   Table1[#All]                 — entire table
//!   Table1[#This Row]            — this row of data
//!   Table1[[#Headers],[Column1]] — header cell of Column1
//!   Table1[[#Totals],[Col1]:[Col3]] — totals row, columns 1-3
//!   Table1[[Col1]:[Col3]]        — data range across columns 1-3

// Re-export parsing functions for tests (accessible via `use super::*` in test modules)
#[cfg(test)]
use compute_parser::{
    find_outer_matching_bracket, is_valid_table_name, parse_bracket_content, parse_structured_ref,
    unescape_column_name,
};

// Re-export types needed by tests (accessible via `use super::*` in test modules)
#[cfg(test)]
pub(crate) use super::types::{
    SpecialItem, StructuredRef, StructuredRefSpecifier, Table, TableColumn, TableRange,
    TableStructureChange,
};

mod adjustment;
mod formatting;
mod resolution;

#[cfg(test)]
mod test_helpers;

#[cfg(test)]
mod tests;

pub use adjustment::*;
pub use formatting::*;
pub use resolution::*;

// ============================================================================
// Row bound helper
// ============================================================================

/// A resolved row range (start inclusive, end inclusive).
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct RowBound {
    pub(crate) start_row: u32,
    pub(crate) end_row: u32,
}

/// A resolved range from structured reference resolution.
///
/// Represents a rectangular region with row bounds and a set of absolute grid column indices.
#[derive(Debug, Clone, PartialEq)]
pub struct ResolvedRange {
    pub start_row: u32,
    pub end_row: u32,
    pub columns: Vec<u32>,
}

/// Result of resolving a structured (table) reference.
/// Contains positional data only — no cell values.
#[derive(Debug, Clone, PartialEq)]
pub struct ResolvedStructuredRef {
    pub sheet: cell_types::SheetId,
    pub ranges: Vec<ResolvedRange>,
}
